### ISSUE TITLE
Roles rewrite

### ISSUE THREAD

**Issue by itsrachelfish** [playasoft/laravel-voldb#28]
The roles system as it stands is very rigid. All roles and permissions are hard coded, which makes it difficult to add new features that change the way event permissions should behave (see issue #20).

The entire system should be rewritten to ensure long term flexibility.

**Comment by itsrachelfish** [comment]
Roles should also have a polymorphic relationship to associate the role with specific events or departments. This would allow "event admins" (issue #30) to only have permissions in certain events and would also allow the creation of "department leads" that only have permissions within specific departments.

**Comment by IzKen** [comment]
I don't think the 2016 VolDB has this type of support. I only see "Staff status" option on users, which makes them an admin.

So I think it only has user & admin level support. Yes, having more granular roles is needed, but lower priority for 2017 I think. For this year we can make those who need to add\edit shift admins I think. This will mean we might have around 30 admins that can screw stuff up, so I would say admin rights would only be given out on an as needed basis and only after they had some "training."

This could be moved to 2018 Milestone, but at least a lower priority for 2017.

**Comment by itsrachelfish** [comment]
I think it's important we get the roles rewrite done ASAP so users can belong to multiple roles simultaneously. Volunteer / Medical / Ranger ... and Uncle Dave was saying that some departments have their own sub-groups.

This would be very helpful for sending notifications to specific groups of people and for more granular permissions down the road. But I agree that issues #20 and #30 aren't necessary for this year.

**Comment by itsrachelfish** [comment]
Todo list:

- [ ] Create roles table
- [ ] Create user_roles table
- [ ] Create event_roles table
- [ ] Convert user role data
- [ ] Delete old role column from users table
- [ ] Convert department roles to shift roles
- [ ] Delete old roles column from departments table
- [ ] Convert shift roles to rows in the new event_roles table
- [ ] Delete old roles column from shifts table
- [ ] Convert schedule roles to rows in the new event_roles table
- [ ] Delete old roles column from schedule table
- [ ] Make sure users still get their appropriate roles
- [ ] Make sure templates / forms are updated to use new roles tables

**Comment by itsrachelfish** [comment]
This is finished as of 7aa444d07d51e22092534b88141cdc02ec173383
